Miscellaneous improvements to the etcd client

giorio94 opened the pull request:
Please review commit by commit, and refer to the corresponding descriptions for additional information.

giorio94 committed:
etcd: remove the waitForInitialSession function

With the refactoring of the lease management logic, which is now
lazily performed on-demand instead of at startup time, it is no
longer necessary to explicitly wait for the initial session to be
established. Hence, let's remove the last usages and the function
itself. Additionally, this makes the behavior of the Update*()
methods consistent with that of the other ones, as well as makes
Close() no longer blocking.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
etcd: rework lease manager to additionally create sessions

Currently, the etcd lease manager is implemented directly leveraging
the Grant() and Keepalive() primitives. Yet, certain functionalities
(i.e., etcd mutexes) depend on the composite Session type. Let's
rework the manager to additionally construct such session objects,
so that in a subsequent commit we can reuse the same logic to replace
the lock session handling. Most notably, making the acquisition of said
session lazy, so that the additional overhead (given the short TTL)
can be avoided when not necessary (e.g., in the clustermesh case, as
agents access remote etcd instances in read-only mode).

The overall acquisition logic is not modified, except for the necessary
adaptations. Unit tests are updated to account for the need to propagate
a full (mocked) client object, instead of the mocked sub-interface only.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
etcd: simplify Close method

Both the client and the lease manager are always guaranteed to be
non-nil. Hence, let's drop the guarding if statements.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
etcd: switch lock session to lease manager

This commit drops the currently used lock session acquisition logic and
replaces it with a new instance of the lease manager, adapting the call
sites. The lock session is still established regardless of whether the
client is used in normal or read-only mode, which will be improved while
subsequently refactoring the status checking logic.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
etcd: drop server version validation

Currently, upon connecting to an etcd cluster, we validate that all its
endpoints have version greater than or equal to 3.1.0. etcd v3.1.0 got
released in 2017, and the currently supported versions [1] are v3.4.x
(first released in 2019) and v3.5.x (first released in 2021). Given the
above and the fact that the check introduces additional load on etcd
during connection establishment (especially relevant in the context
of large scale cluster meshes), let's drop it.

[1]: https://etcd.io/docs/v3.5/op-guide/versioning/

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

giorio94 committed:
kvstore: explicitly wait for dummy client to connect

The SetupDummy() function used to setup a kvstore client for testing
purposes (indirectly) spawns a goroutine which panics if the connection
to the kvstore fails. Yet, this is prone to race conditions, as in fast
tests we might close the client before that the connection check has
completed, causing the error to be raised.

To avoid this issue, let's explicitly wait for the connection to be
established during the setup. We never observed this race condition
in the past because the Close() operation used to block until the
connection was established (but that could hang in case of errors).

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

tklauser reviewed (approved):
LGTM, but it's been a while since I last touched that part of the code base. I've requested a review from @marseel per https://github.com/cilium/cilium/pull/28834#issuecomment-1782950554

marseel commented on pkg/kvstore/etcd.go:
```diff
@@ -822,59 +803,13 @@ func makeSessionName(sessionPrefix string, opts *ExtraOptions) string {
-// checkMinVersion checks the minimal version running on etcd cluster.  This
-// function should be run whenever the etcd client is connected for the first
-// time and whenever the session is renewed.
-func (e *etcdClient) checkMinVersion(ctx context.Context, timeout time.Duration) error {
```
It definitely makes sense to remove version validation here. Maybe we could follow-up in cilium-cli to get Etcd version and print some warning in case it's old/unsupported version? While it doesn't matter for clustermesh too much, for kvstore mode it could make sense. If it's fine with you, maybe let's create follow-up issue in cilium-cli?

giorio94 replied:
:thinking: Are you thinking about checking that at installation time or when doing `cilium status`? The main issue I see is that it is tricky to ensure that the CLI is always able to access the etcd members, independently of whether they are outside or inside the cluster. And I'm not 100% sure whether the effort is really worth considering that Cilium is currently compatible with any supported version of etcd.

marseel replied:
I was thinking about doing that in `cilium status`. For now, just creating issue in cilium-cli to capture that we might want to introduce this check. Even though we support all versions, I think it would be nice to suggest upgrades etc in case users are running Etcd older than 3.4, or 3.5 with [known data inconsistency issues](https://groups.google.com/a/kubernetes.io/g/dev/c/B7gJs88XtQc/m/rSgNOzV2BwAJ) (versions 3.5.0-3.5.2)

giorio94 replied:
https://github.com/cilium/cilium-cli/issues/2107

marseel commented on pkg/kvstore/etcd_lease.go:
```diff
@@ -69,12 +63,26 @@ func newEtcdLeaseManager(cl etcdLeaseClient, ttl time.Duration, limit uint32, ex
+	return session.Lease(), nil
+}
+
+// GetLeaseID returns a session, and associates it to the given key. It leverages
```
nit: s/GetLeaseID/GetSession/

marseel commented on pkg/kvstore/etcd_lease.go:
```diff
@@ -199,57 +209,62 @@ func (elm *etcdLeaseManager) Wait() {
 }
 
-func (elm *etcdLeaseManager) newLease(ctx context.Context) (c client.LeaseID, cancelFn context.CancelFunc, err error) {
+func (elm *etcdLeaseManager) newLease(ctx context.Context) (session *concurrency.Session, err error) {
```
nit: newLease/newSession

marseel commented on pkg/kvstore/etcd.go:
```diff
@@ -359,11 +352,13 @@ type etcdClient struct {
 
+	// leaseManager manages the lazy acquisition of etcd leases for generic purposes
 	leaseManager *etcdLeaseManager
+	// lockLeaseManager manages the lazy acquisition of etcd leases for locking
```
I would probably remove `lazy` from the comments here as it's implementation detail, not important in current context imho.

marseel reviewed (commented):
Looks good, I will go over this PR one more time to make sure I checked all edge cases.

marseel commented on pkg/kvstore/etcd_lease.go:
```diff
@@ -94,7 +102,7 @@ func (elm *etcdLeaseManager) GetLeaseID(ctx context.Context, key string) (client
 			elm.mu.Unlock()
 
-			return elm.current, nil
+			return info.session, nil
```
We should check here if sessions in not Orphaned I think.
in https://github.com/cilium/cilium/pull/28834/files#diff-85b5ad3b5a6b7004aafad37cdbb264e395fd491231f4c8cae8ce88554e7d1145R193
we can Orphan session, but it potentially can still be in `elm.leases` as we release lock after that and remove it later on in `waitForExpiration`

marseel replied:
I was thinking about something like this, to make it less likely to happen:
```
select {
case <-info.session.Done():
    continue
default:
}
```

giorio94 replied:
The main issue is that we cannot completely get rid of the uncertainty period, due to the delay between server-side expiration and client-side detection (with consequent removal of the session from the list of known ones). While additional checks could help slightly reducing that window, they also make the logic more complex (as they would break the assumption that sessions get GCed only by the `waitForExpiration` function). For this reason, the idea here is to essentially rely on client-side retries in the unlikely case that the session is expired. 

As discussed offline, I've extended the comments of the GetLease/GetSession functions for additional clarity about that assumption: https://github.com/cilium/cilium/pull/28834/commits/48518d6781c67198f62fbd0db094c820309bf9cf

giorio94 committed:
etcd: extend GetLease/GetSession comments for additional clarity

Let's explicit the fact that there's a small possibility that the
returned lease/session is already expired, or gets expired immediately
before use, due to the time window between server-side lease expiration
and client-side detection. As we cannot completely remove this
uncertainty period, let's just adopt the simplest approach here, and
rely on retries to eventually converge.

Signed-off-by: Marco Iorio <marco.iorio@isovalent.com>

marseel reviewed (approved):
Looks good, thanks!
